### PR TITLE
[core] Add the concept of wire sets.

### PR DESCRIPTION
**Pull request by schweitzpgi**
A wire set is a new abstraction in Quake. A wire set is a finite set of wires (a.k.a. qubits). Each wire has a unique identity. All wires remain linear types.

The usage model is to borrow a wire from a set, use it, and return it to the set.

A wire set can also model other physical properties of qubits such as topology via a sparse adjecency map (for copious generality, including unidirectional links).

~Requires #1863 - the diffs will be clearer once the referenced PR is merged.~

### CODE REVIEW

**Comment by schweitzpgi** [comment]
ping: needs review to unblock the interns

File: include/cudaq/Optimizer/Dialect/Quake/QuakeOps.td [hunk 1]
```diff
@@ -546,6 +546,89 @@ def quake_SinkOp : QuakeOp<"sink"> {
...
+    ```
+  }];
+
+  let arguments = (ins
+    FlatSymbolRefAttr:$set_name,
+    I32Attr:$identity
+  );
+  let results = (outs WireType);
+  
+  //let hasVerifier = 1;
```
Comment (bmhowe23) [inline include/cudaq/Optimizer/Dialect/Quake/QuakeOps.td:612]: Why is this commented out?

Comment (schweitzpgi) [reply include/cudaq/Optimizer/Dialect/Quake/QuakeOps.td]: Added with 38d0b76

**Review by bmhowe23 — APPROVED** [review]
LGTM

**Pull request merged**